### DISCUSSION TITLE
Existing records are altered unintentionally when commiting new ones

### DISCUSSION THREAD

**Discussion by aalvrz** [sqlalchemy/sqlalchemy#7068] [Usage Questions] [answered]
### Describe the bug

I am not sure if this is a MySQL bug or SQLAlchemy bug but I figured I would try reporting this here.

In my unit tests I am instantiating and committing some objects from a model and I am passing an already persisted `team` object as relationship on instantiation. However each subsequent creation sets the previous entry's `team` (or `team_id` FK) 
to NULL for some unknown reason!

### To Reproduce

```python
site_1 = SiteFactory(team=team)
sa_1 = SiteAccess(site=site_1, team=team, permission=1)
session.add(sa_1)
session.commit()
```

This creates the `SiteAccess` entry without any problems, but if I create another one with another site but same team, the `team_id` column from the record above (`sa_1`) is set to NULL 🤯 🤯 🤯 :

```python
site_2 = SiteFactory(team=team)
sa_2 = SiteAccess(site=site_2, team=team, permission=1)
session.add(sa_2)
session.commit()
``` 

```
|id |site_id|user_id|team_id|permission|expiration_date|created_at         |is_deleted|
|---|-------|-------|-------|----------|---------------|-------------------|----------|
|1  |2593   |       |       |1         |               |2021-09-22 18:02:00|0         |
|2  |2594   |       |367    |1         |               |2021-09-22 18:02:21|0         |
```

### Error

No errors

### Versions

- OS: Ubuntu 21.04
- Python: 3.9.5
- SQLAlchemy: 1.3.18
- Database: MySQL 5.7
- DBAPI: PyMySQL==0.10.0

### Additional context

Here is the model declaration:

```python
class SiteAccess(Base, DeletableModel):
    __tablename__ = "site_access"
    __table_args__ = (
        UniqueConstraint("site_id", "user_id"),
        UniqueConstraint("site_id", "team_id"),
    )

    id = Column(INTEGER(10), primary_key=True)
    site_id = Column(ForeignKey("company.company_id"), nullable=False, index=True)
    user_id = Column(ForeignKey("users.user_id"))
    team_id = Column(ForeignKey("team.group_id"))
    permission = Column(TINYINT(1), nullable=False)
    expiration_date = Column(DateTime)

    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)

    site = relationship("Site")
    user = relationship("User", backref=backref("site_access", uselist=False))
    team = relationship("Team", backref=backref("site_access", uselist=False))
```

I don't know if there is something wrong with the declaration to allow something like this to happen.

Here is the DDL:

```sql
CREATE TABLE `site_access` (
  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
  `site_id` int(10) unsigned NOT NULL,
  `user_id` int(10) unsigned DEFAULT NULL,
  `team_id` int(10) unsigned DEFAULT NULL,
  `permission` tinyint(3) unsigned NOT NULL DEFAULT '0',
  `expiration_date` datetime DEFAULT NULL,
  `created_at` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP,
  `is_deleted` tinyint(1) NOT NULL DEFAULT '0',
  PRIMARY KEY (`id`),
  UNIQUE KEY `unique_site_and_user_ids` (`site_id`,`user_id`),
  UNIQUE KEY `unique_site_and_team_ids` (`site_id`,`team_id`),
  KEY `user_id` (`user_id`),
  KEY `team_id` (`team_id`),
  CONSTRAINT `fk_site_access_siteid` FOREIGN KEY (`site_id`) REFERENCES `company` (`company_id`) ON DELETE CASCADE,
  CONSTRAINT `site_access_ibfk_2` FOREIGN KEY (`user_id`) REFERENCES `users` (`user_id`),
  CONSTRAINT `site_access_ibfk_3` FOREIGN KEY (`team_id`) REFERENCES `team` (`group_id`)
) ENGINE=InnoDB AUTO_INCREMENT=48 DEFAULT CHARSET=utf8
```

**Comment by zzzeek** [comment, marked as answer]
hey there - 

your issue is caused by this:

     team = relationship("Team", backref=backref("site_access", uselist=False))

lets step through:

1. set sa_1 with team t1:
    
    sa_1 = SiteAccess(..., team=t1)

2. set sa_2 with the *same* team t1:

   sa_2 = SiteAccess(..., team=t1)

3. now per your mapping, you have a backref set up, "Team.site_access", with uselist is False.     So what do you expect to see when you do this?
   
    some_sa = t1.site_access

What's the value of "some_sa"  ?   is it "sa_2"?  or "sa_1"?    it can't be both, because you said uselist=False.   So when you set "sa_2 = SiteAccess( ... team = t1)" , the ORM had no choice but to remove the association of "t1" from "sa_1" otherwise it can't fulfill the backref.

the fix is to take off that uselist=False, Team.site_access has to be a collection if you expect more than one SiteAccess object to refer to it at the same time.

**Reply by aalvrz** [reply]
@zzzeek You are correct. Thanks for the detailed comment. Should an error or warning be raised instead though? Seems like it could be easy for this to go unnoticed.

**Reply by zzzeek** [reply]
the uselist=False use case you have there is known as "one-to-one", it's a real way to configure things, so there's nothing to warn.
